keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+
+## Proposal
+
+In order to support dual-stack in Kubernetes clusters, Kubernetes needs to have awareness of and support dual-stack addresses for pods, nodes and services. To make this happen, the following changes will be required:
+
+- Kubernetes needs to be made aware of multiple IPs per pod (tested with up to one IPv4 and up to two IPv6 addresses).
+- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
+- Kubernetes needs to be configurable for up to two service CIDRs.
+- Backend pods for a service can be dual stack. For the first release of dual-stack support, each IPv4/IPv6 address of a backend pod will be treated as a separate Kubernetes endpoint.
+- Kube-proxy needs to support IPv4 and IPv6 services in parallel (e.g. drive iptables and ip6tables in parallel).
+- Health/liveness/readiness probes should work for dual-stack pods. For the first release of dual-stack support, no additional configuration "knobs" will be added for probe definitions, and a probe is deemed successful if either an IPv4 or IPv6 response is received
+- Kubectl commands and output displays will need to be modified for dual-stack.
+- Kubeadm support will need to be added to enable spin-up of dual-stack clusters.
+- New e2e test cases will need to be added to test parallel IPv4/IPv6 connectivity between pods, nodes, and services.
+
+### Awareness of Multiple IPs per Pod
+
+Since Kubernetes Version 1.9, Kubernetes users have had the capability to use dual-stack-capable CNI network plugins (e.g. Bridge + Host Local, Calico, etc.), using the 
+[0.3.1 version of the CNI Networking Plugin API](https://github.com/containernetworking/cni/blob/spec-v0.3.1/SPEC.md), to configure multiple IPv4/IPv6 addresses on pods. However, Kubernetes currently captures and uses only the first IP address in the list of assigned pod IPs that a CNI plugin returns to Kubelet in the [CNI Results structure](https://github.com/containernetworking/cni/blob/spec-v0.3.1/SPEC.md#result).
+
+This proposal aims to extend the Kubernetes Pod Status API so that Kubernetes can track and make use of multiple IPv4/IPv6 address assignments per pod.
+
+#### Versioned API Change: PodStatus v1 core
+In order to maintain backwards compatibility for the core V1 API, this proposal adds a new "ExtraPodIPs" field to the core V1 version of the [PodStatus V1 core API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#podstatus-v1-core), while maintaining the existing "PodIP" field. This change is in accordance with the [Kubernetes API change
```

Comment:
for context, maybe include an appendix or footnote detailing this?

The issue with compat here is older clients.

First: We can't really stop populating this field, or we break clients who don't understand `PodIPs`.

A simpler-seeming answer might be to have both `PodIP` and `PodIPs` where the plural always includes the singular.  Again, the problem is older clients (think about a failed kubelet update which gets rolled back).

Consider an existing pod with `PodIP` = "X" and `PodIPs" = [ "X", "Y" ].  If we have a client that only knows `PodIP`, and it updates that field to "Z", but does not update `PodIPs`, the inclusivity invariant is violated.

If we make it optionally inclusive, we still may have muddled the intention - the client changed "X" to "Z", did they mean that "X" is no longer preferred or no longer present at all?

The same problem exists in the other direction. If the older client updated `PodIP` from "Z" to "X", that collides with `PodIPs`.  Did the user mean the same "X"?

That doesn't leave much room for viable solutions.  It can't be guaranteed exclusive nor can it be guaranteed inclusive, and optionally inclusive is confusing at best.  The only real option here is to consider the real-life chance of detriment.  Pod IPs *do* change (e.g. if the node reboots quickly, it can come back without evicting the pod, but IPAM can choose a new IP), but it's not expected to be common.  It's also uncommon for multiple clients to be writing to Pod.status, so we really only need to consider the upgrade/rollback case for mutations.

Thus, I think it's reasonably safe to say it is optionally inclusive, and to define the edge case behavior.

This might be worth having a very focused sig-architecture chat about.  There may be better options that I haven't thought about or precedents that I don't know.

Other options, off the top of my head:

* Linked fields.  Any update to singular with a corresponding update to plural can validate inclusivity/exclusivity.  Any update to singular without an update to plural forces plural to some defined state (inclusive or exclusive TBD).  Any update to plural without an update to singular forces singular to some defined state.

* move Pod to v2, make plural be read-only in v1 and collaps singular and plural in v2.  Changes through v2 extract the first item (or some other heuristic) into singular in v1 view.

Reply:
> There may be better options that I haven't thought about or precedents that I don't know.

The "[API changes](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#on-compatibility)" doc explicitly recommends this approach. (The example with `Param` and `ExtraParams`.)

Reply:
Ah, that makes sense. I hadn't thought about kubelet downgrades (scary!).

I would really like to attach metadata to IPs - thoughts about how?

Reply:
I think we need it too.  This is a pattern (pluralising a field) that doesn't pop up often enough to have a pattern that I know of, so maybe we will have to trailblaze a bit.

Reply:
Are there two different suggestions here? One being that the old field has the first (primary) value and the array has additional entries (seems like the API changes doc specifies this). The other being the new array containing the entry held in the original field (duplicating)? Just trying to get this straight.

Would it be good to have a separate section to discuss field extension in general, the pros, cons, corner cases, etc., since we have several of these?

Reply:
@pmichali - Right, the two *main* options being discussed could be called exclusive, where you have 
```
podIP="X"
podIPs=["Y", "Z"]
```
versus inclusive, where you have:
```
podIP="X"
podIPs=["X", "Y", "Z"]
```
The exclusive is what was originally proposed, and is based on the pluralized example in the [API changes doc](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#on-compatibility). And then there's an "optional inclusive" variant that seems to be a inclusive/exclusive hybrid (old clients only update the singular podIP, so podIPs inclusivity is lost upon modification of podIP). I had been thinking that either pure exclusive or pure inclusive models should work - old clients remain agnostic of the pluralized field, but the conversion logic in the server (stuff that converts from versioned APIs to internal versions, and vice versa) is smart enough to maintain coherency in the podIP and podIPs fields... but I could be missing nuances.

Yes, the "extra section" is what @thockin is suggesting (in appendix or footnote form), although I'm thinking that if the  [API changes doc](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#on-compatibility) can be updated accordingly, we could just point to that.

Reply:
@leblancd Not everyone will do a read-modify-write, and even if they do ... older clients will drop unknown fields AND they can send a patch.  So pathologically:

Mandatory exclusive:
* podIP="X", podIPs=["Y", "Z"]
* client sends a patch operation changing podIP to "Z".
=> If we fail that at the apiserver, the client has no idea why an update that used to work doesn't any more.  If we don't fail it, we can either remove "Z" from podIPs (mutating user input is generally verboten) or we can ignore it and break the invariant.

Mandatory inclusive:
* podIP="X", podIPs=["X", "Y"]
* client sends a patch operation changing podIP to "Z".
=> If we fail that at the apiserver, the client has no idea why an update that used to work doesn't any more.  If we don't fail it, we can either add  "Z" to podIPs (mutating user input is generally verboten) and remove "X" (loses user metadata) or we can ignore it and break the invariant.

I'm saying we need to find a cleaner way and set precedent very carefully.  Optionally-inclusive (which I guess is == optionally exclusive :) is still confusing.
* podIP="X", podIPs=["X", "Y"]
* client sends a patch operation changing podIP to "Z".
=> Is the "X" in podIPs valid any more?

I want to consider more clever ideas (if anyone has some).  This *will* happen again.

Reply:
@thockin It's not clear to me how we can solve this.

We talked elsewhere in this review about including metadata (e.g. annotations) along with IPs in the podIPs/extraPodIPs array. If we do include metadata, I think we'd be better off (or it would at least be simpler) using an inclusive model. The reason being that the singular podIP field needs to remain a simple IP address (the "default" IP address) without metadata. If we use an inclusive model, we can store metadata associated with the default IP address in the podIPs array. (For exclusive, we'd have to add yet another field for that IP address' metadata).

Of course, if an old client changes podIP, the old client is agnostic to metadata, so the new podIP will be added without metadata. The podIPs field would then have a mix of entries with/without metadata. I don't know how bad of a problem this would be, but there's probably no way around it.

For the mandatory inclusive scenario, I would consider the "X" in podIP and the "X" in podIPs to be duplicate copies of the same field, so I would think that if client replaces/overwrites the "X" in podIP with "Z", then "Z" should be added to podIPs, and the (duplicate) "X" in podIPs should be removed, along with its metadata, IMO. Is mutating user input so bad that we shouldn't even consider it?

I'm looking for guidance. In the meantime, I'll rewrite this section assuming a mandatory inclusive model for now, because of the metadata concerns I mentioned above.

Reply:
@leblancd I think it is about harm reduction.  If someone uses an older client, it can't stop working.  In practice I doubt we will have many problems with multi-writer, so I think we can opt for the optionally-inclusive mode.

I've put this on the agenda for sig-arch this Thursday, though we have a pretty full list, so it may push to next week.

https://docs.google.com/document/d/1BlmHq5uPyBUDlppYqAAzslVbAO8hilgjqZUTaNXUhKM/edit?ts=5b43a34a#